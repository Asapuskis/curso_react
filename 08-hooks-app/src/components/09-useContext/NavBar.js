import React from 'react';
import { NavLink } from 'react-router-dom';

// Creamos un componente para hacer un menú de navegación
export const NavBar = () => {
    return (
        // Nav simple sin clases de bootstrap
        // El componente NavLink sirve para hacer las navegaciones sin tener que recargar toda la página. to indica a qué página.
        // <nav>
        //     <ul>
        //         <li>
        //             <NavLink to="/">Home</NavLink>
        //         </li>
        //         <li>
        //             <NavLink to="/about">About</NavLink>
        //         </li>
        //         <li>
        //             <NavLink to="/login">Login</NavLink>
        //         </li>
        //     </ul>
        // </nav>
        <nav className="navbar navbar-expand-sm navbar-dark bg-dark">
            <div className="container-fluid">
                <NavLink className="navbar-brand" to="/">useContext App</NavLink>

                <div className="collapse navbar-collapse" id="navbarNavAltMarkup">
                    <div className="navbar-nav">
                        {/* Si añadimos exact activeClassName='active' cuando la ruta del navegador coincida con el NavLink, se marcará como activo
                        en el NavBar. OJO, a partir de la versión 6 de react-router el componente NavLink no tiene activeClassName!! */}
                        <NavLink exact='true' activeClassName='active' className="nav-link" aria-current="page" to="/">Home</NavLink>
                        <NavLink exact='true' activeClassName='active' className="nav-link" to="/about">About</NavLink>
                        <NavLink exact='true' activeClassName='active' className="nav-link" to="/login">Login</NavLink>
                    </div>
                </div>
            </div>
        </nav>
    )
}
