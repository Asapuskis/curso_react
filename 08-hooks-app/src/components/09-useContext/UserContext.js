import { createContext } from 'react';

// createContext genera un objeto context que exportado, podemos acceder a Ã©l desde cualquier lado.
export const UserContext = createContext(null);