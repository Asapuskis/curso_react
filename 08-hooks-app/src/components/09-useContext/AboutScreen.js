import React, { useContext } from 'react'
import { UserContext } from './UserContext'

export const AboutScreen = () => {

    const {user, setUser} = useContext(UserContext);

    const handleClick = () => {
        setUser( {} );
    }

    return (
        <>
            <h1>AboutScreen</h1>
            <hr/>
            <pre>
                { JSON.stringify( user, null, 3)}
            </pre>
            {/* Metemos un logout para demostrar que podemos utilizar funciones pasadas con el context y que funciona de lux. */}
            <button
                className='btn btn-warning'
                onClick={ handleClick }>
                Logout
            </button>
        </>
    )
}
