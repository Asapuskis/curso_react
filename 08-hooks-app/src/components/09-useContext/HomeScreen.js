import React, { useContext } from 'react'
import { UserContext } from './UserContext'

export const HomeScreen = () => {

    // Mediante la función useContext podemos recoger un Context creado y que esté en el scope del componente.
    // En este caso, cogemos el context UserContext creado en UserContext.js
    const userContext = useContext(UserContext)

    return (
        <>
            <h1>HomeScreen</h1>
            <hr/>
            <pre>
                { JSON.stringify( userContext.user, null, 3) }
            </pre>
        </>
    )
}
