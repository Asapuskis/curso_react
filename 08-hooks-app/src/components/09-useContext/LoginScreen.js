import React, { useContext } from 'react'
import { UserContext } from './UserContext'

export const LoginScreen = () => {

    const { setUser } = useContext(UserContext);

    return (
        <>
            <h1>LoginScreen</h1>
            <hr/>
            {/* AÃ±adimos un user para demostrar que podemos utilizar las funciones pasadas por context */}
            <button
                className='btn btn-primary'
                onClick={ () => setUser( {
                    id: 123,
                    name: 'Paco'
                })}>
                Login
            </button>
        </>
    )
}
