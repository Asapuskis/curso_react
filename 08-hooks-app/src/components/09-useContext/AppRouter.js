import React from 'react'
import {
    BrowserRouter as Router,
    Routes,
    Route,
  } from "react-router-dom";
import { AboutScreen } from './AboutScreen';
import { HomeScreen } from './HomeScreen';
import { LoginScreen } from './LoginScreen';
import { NavBar } from './NavBar';

// Componente para hacer la navegación entre páginas o componentes.
export const AppRouter = () => {
    return (
        // El componente Router es el componente padre donde tienen que encontrarse los demás componentes de navegación
        <Router>
            <div>
                <NavBar />
                <div className='container'>
                    {/* Routes hace de switch, es decir, va a evaluar la url y dependiendo del valor, navegará a uno u otro componente. */}
                    <Routes>
                        {/* path es la url que hay que añadir en el navegador, si se añade exact, implica que el path tiene que ser exacto, si no se pusiese
                        evaluaría mediante una expresión regular y daría fallos
                            element indica el componente o página a la que va a navegar. */}
                        <Route exact='true' path='/' element={ <HomeScreen /> }/>
                        <Route exact='true' path='/about' element={ <AboutScreen /> }/>
                        <Route exact='true' path='/login' element={ <LoginScreen /> }/>
                        {/* Path por defecto si no encuentra ninguna ruta, recomendado crear un 404 o algo así para handlear este tipo de excepciones */}
                        <Route element={ <HomeScreen /> }/>
                    </Routes>
                </div>
            </div>
        </Router>
    )
}
