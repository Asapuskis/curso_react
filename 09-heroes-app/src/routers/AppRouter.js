import React from 'react';
import { Routes, Route, BrowserRouter } from 'react-router-dom';
import { LoginScreen } from '../components/login/LoginScreen';
import { DashboardRoutes } from './DashboardRoutes';
import { PrivateRoute } from './PrivateRoute';
import { PublicRoute } from './PublicRoute';

// Router para la web cuando se ha logeado un usuario.
export const AppRouter = () => {
  return (
    <BrowserRouter>
        <Routes>
            {/* Ruta al login */}
            <Route path="/login" element={
              // En este caso, solo ser치 publico LoginScreen si. Se controla en el PublicRoute.js
              <PublicRoute>
                <LoginScreen />
              </PublicRoute>
            } />
            {/* Ruta para manejar el resto de rutas mediante el componente <DashboardRoutes />
              Ir치 dentro de PrivateRoute que ser치 quien verifique si el usuario est치 autorizado para ver las rutas de DashboardRoutes o no.*/}
            <Route path="/*" element={
              <PrivateRoute>
                <DashboardRoutes />
              </PrivateRoute>
            } />
        </Routes>
    </BrowserRouter>
  )
}
