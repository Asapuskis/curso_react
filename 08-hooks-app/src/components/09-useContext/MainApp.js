import React, { useState } from 'react'
import { AppRouter } from './AppRouter'
import { UserContext } from './UserContext'

export const MainApp = () => {

    const [user, setUser] = useState({});

    // Objeto que pasaremos con el provider a los componentes hijos.
    const valueProvide = {
        user,
        setUser
    }

    // Si importamos el UserContext y le agregamos el provider, lo hacemos accesible a todos los componentes hijo que englobe.
    // El Provider necesita un prop llamado value que indica qué es lo que se va a compartir a los demás componentes.
    return (
        <UserContext.Provider value={valueProvide}>
            <AppRouter />
        </UserContext.Provider>
    )
}
