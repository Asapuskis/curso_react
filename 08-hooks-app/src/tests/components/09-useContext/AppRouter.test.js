import { mount } from 'enzyme';
import { AppRouter } from '../../../components/09-useContext/AppRouter';
import { UserContext } from '../../../components/09-useContext/UserContext';
import { user } from '../../fixtures/user';


describe('Pruebas de <AppRouter />', () => {

    const wrapper = mount(
        <UserContext.Provider value={user}>
            <AppRouter />
        </UserContext.Provider>
    );

    test('Debe mostrarse un snapshot correctamente', () => {
        expect( wrapper ).toMatchSnapshot();
    });

})