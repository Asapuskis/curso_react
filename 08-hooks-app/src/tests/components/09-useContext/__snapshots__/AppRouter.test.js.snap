// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas de <AppRouter /> Debe mostrarse un snapshot correctamente 1`] = `
<div>
  <nav
    className="navbar navbar-expand-sm navbar-dark bg-dark"
  >
    <div
      className="container-fluid"
    >
      <a
        aria-current="page"
        className="navbar-brand active"
        href="/"
        onClick={[Function]}
      >
        useContext App
      </a>
      <div
        className="collapse navbar-collapse"
        id="navbarNavAltMarkup"
      >
        <div
          className="navbar-nav"
        >
          <a
            activeClassName="active"
            aria-current="page"
            className="nav-link active"
            exact="true"
            href="/"
            onClick={[Function]}
          >
            Home
          </a>
          <a
            activeClassName="active"
            className="nav-link"
            exact="true"
            href="/about"
            onClick={[Function]}
          >
            About
          </a>
          <a
            activeClassName="active"
            className="nav-link"
            exact="true"
            href="/login"
            onClick={[Function]}
          >
            Login
          </a>
        </div>
      </div>
    </div>
  </nav>
  <div
    className="container"
  >
    Array [
      <h1>
        HomeScreen
      </h1>,
      <hr />,
      <pre />,
    ]
  </div>
</div>
`;
